### ISSUE TITLE
Right mouse clicks sometimes don't work

### ISSUE THREAD

**Issue by 3hhh** [qtile/qtile#5067]
### Issue description

I noticed that in some windows right mouse clicks stop working after some time.

Usually it happens in `firefox` windows on the web page (sometimes the address bar can be right clicked, but the web page cannot; `chromium` seems to be more resilient) and usually after a qtile restart via `mod-r`, `systemctl restart lightdm`, window minimization and re-maximization or a screen change (e.g. laptop dock/undock). The latter seems to cause a ton of other issues as well, but I might create a separate bug for that in the future.

### Version

0.29.0

### Backend

X11 (default)

### Required

- [X] I have searched past issues to see if this bug has already been reported, and it hasn't been.
- [X] I understand that people give their precious time for free, and thus I've done my very best to make this problem as easy as possible to investigate.

**Comment by 3hhh** [comment]
I managed to reproduce this issue on awesomeWM, so that I think it's [an issue with my OS](https://github.com/QubesOS/qubes-issues/issues/9507) rather than qtile. On awesomeWM it's only reproducible on dock/undock though, not on awesomeWM restarts.

So I'll close this one for now.

**Comment by tych0** [comment]
This sounds vaguely like something about the focus being broken, similar to 7b43c9f8af839831306a99d717a9afb0280a7189. Since you can reproduce with awesome you're right that it's probably qubes, but poking at the RevertTo behavior of the applications you're seeing it with might yield interesting things.

**Comment by 3hhh** [comment]
So now I realized that there's a bug in qtile as well:
The right click context menu windows of several applications (e.g. `firefox` & `thunderbird`) are sometimes drawn _behind_ the current window, effectively making them invisible. This may be specific to MonadTall (I haven't tested other layouts).

I can reproduce this with the default qtile config. It requires opening firefox and doing a lot of workspace switches, adding other windows, moving it around, enable/disable floating, ...

However it's harder to make screenshots with the default config, so here's a screenshot with my config:
![Screenshot_2024-11-16_10-52-56](https://github.com/user-attachments/assets/1577a18e-d780-4014-9f54-3f9db8757860)

(The context menu should start above the firefox window, not below.)

**Comment by 3hhh** [comment]
A few more notes:
- Once the issue happens, it persists at least until a `firefox` restart.
- I noticed that setting `widget.gtk.native-context-menus` to `true` in `about:config` appears to not cause the issue (the context menu is always shown). However then menu left clicks appear to have no effect (i.e. the menu draws, but has no funtionality) and I receive the error:
```
(firefox:4688): Gtk-WARNING **: 20:45:40.203: no trigger event for menu popup
[Parent 4688, Main Thread] WARNING: no trigger event for menu popup: 'glib warning', file /builds/worker/checkouts/gecko/toolkit/xre/nsSigHandlers.cpp:187
```

**Comment by 3hhh** [comment]
Non-working firefox context menu:
```
WM_HINTS(WM_HINTS):
                Client accepts input or input focus: True
                Initial state is Normal State.
                window id # of group leader: 0xa00001
WM_WINDOW_ROLE(STRING) = "Popup"
_NET_WM_BYPASS_COMPOSITOR(CARDINAL) = 2
XdndAware(ATOM) = BITMAP
_NET_WM_OPAQUE_REGION(CARDINAL) = 
WM_TRANSIENT_FOR(WINDOW): window id # 0xa0002b
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_POPUP_MENU
_NET_WM_SYNC_REQUEST_COUNTER(CARDINAL) = 10485853, 10485854
_NET_WM_USER_TIME(CARDINAL) = 93713
_NET_WM_USER_TIME_WINDOW(WINDOW): window id # 0xa0005c
WM_CLIENT_LEADER(WINDOW): window id # 0xa00001
_NET_WM_PID(CARDINAL) = 1591
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "disp6640"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified location: 0, 0
                program specified minimum size: 0 by 0
                program specified base size: 0 by 0
                window gravity: NorthWest
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING, _NET_WM_SYNC_REQUEST
WM_CLASS(STRING) = "Firefox", "firefox"
WM_ICON_NAME(STRING) = "Firefox"
_NET_WM_ICON_NAME(UTF8_STRING) = "Firefox"
WM_NAME(STRING) = "Firefox"
_NET_WM_NAME(UTF8_STRING) = "Firefox"
```

Working chromium context menu:
```
_NET_WM_STATE(ATOM) = _NET_WM_STATE_ABOVE
_NET_WM_USER_TIME(CARDINAL) = 0
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified location: 1276, 602
                window gravity: Static
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x2, 0x0, 0x0, 0x0, 0x0
XdndAware(ATOM) = BITMAP
_NET_WM_ICON(CARDINAL) =        Icon (128 x 128):
_NET_WM_BYPASS_COMPOSITOR(CARDINAL) = 2
WM_NAME(UTF8_STRING) = 
_NET_WM_NAME(UTF8_STRING) = 
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_MENU
_NET_WM_PID(CARDINAL) = 2003
WM_CLIENT_MACHINE(STRING) = "disp6640"
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, _NET_WM_PING, _NET_WM_SYNC_REQUEST
```

**Comment by tych0** [comment]
Thanks for the xprop, I think we do have a bug: https://github.com/tych0/qtile/commit/0c7f945f8dcba3e2a602bf73267fd285beb68822

Can you try that patch and see if it fixes things for you?

**Comment by tych0** [comment]
Hmm, ok. seems there is a bug in [change_layer()](https://github.com/qtile/qtile/blob/02e92118928f851dd27e3a4f8d9205cec12238bc/libqtile/backend/x11/window.py#L992-L1203) somewhere then too. I don't understand this function at all, maybe cc @elParaguayo sees something obvious? The difference here is the presence of `WM_TRANSIENT_FOR`.

**Comment by elParaguayo** [comment]
Long shot... can you check if #4583 solves this?

**Comment by 3hhh** [comment]
> Long shot... can you check if #4583 solves this?

No, unfortunately not either.

**Comment by tych0** [comment]
Oh, interesting. So that makes sense as to why my change didn't work: it's not a new window. But I wonder: if we're placing windows with transient_for(), seems like they should always go on top? Does https://github.com/tych0/qtile/commit/c300d5191c0efdd3132aeb844c0764918b8695b4 help?

**Comment by elParaguayo** [comment]
I thought I had some code to move transient_for windows with their parent...

**Comment by 3hhh** [comment]
> Does [tych0@c300d51](https://github.com/tych0/qtile/commit/c300d5191c0efdd3132aeb844c0764918b8695b4) help?

No, it remains as it was:
![Screenshot_2024-11-18_19-51-12](https://github.com/user-attachments/assets/9fad49b5-031e-4e61-a5e2-bd2534975fd6)

Btw as can be seen from the screenshot it may be hidden behind windows on the right, which effectively makes it invisible on the right side.

I even tried to patch out my anti-focus-steal code (which however only runs for new windows, i.e. is unlikely to be in the way) and set `focus_on_window_activation = "always"`, but to no avail.

**Comment by 3hhh** [comment]
Btw from the below thunderbird examples I'm not entirely convinced it's the transient setting:

thunderbird gtk context menu (working fine):
```
WM_HINTS(WM_HINTS):
                Client accepts input or input focus: True
                Initial state is Normal State.
                window id # of group leader: 0x800001
_NET_WM_SYNC_REQUEST_COUNTER(CARDINAL) = 8389288, 8389289
_NET_WM_USER_TIME(CARDINAL) = 819470
WM_CLIENT_LEADER(WINDOW): window id # 0x800001
_NET_WM_PID(CARDINAL) = 6193
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "e-mail"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified size: 10 by 10
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING, _NET_WM_SYNC_REQUEST
WM_ICON_NAME(STRING) = "Thunderbird"
_NET_WM_ICON_NAME(UTF8_STRING) = "Thunderbird"
WM_NAME(STRING) = "Thunderbird"
_NET_WM_NAME(UTF8_STRING) = "Thunderbird"
WM_HINTS(WM_HINTS):
                Client accepts input or input focus: True
                Initial state is Normal State.
                window id # of group leader: 0x800001
WM_TRANSIENT_FOR(WINDOW): window id # 0x800092
XdndAware(ATOM) = BITMAP
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_POPUP_MENU
_NET_WM_SYNC_REQUEST_COUNTER(CARDINAL) = 8389292, 8389293
_NET_WM_USER_TIME(CARDINAL) = 819470
_NET_WM_USER_TIME_WINDOW(WINDOW): window id # 0x8002ab
WM_CLIENT_LEADER(WINDOW): window id # 0x800001
_NET_WM_PID(CARDINAL) = 6193
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "e-mail"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified minimum size: 224 by 498
                program specified maximum size: 224 by 498
                program specified base size: 0 by 0
                window gravity: NorthWest
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING, _NET_WM_SYNC_REQUEST
WM_CLASS(STRING) = "thunderbird", "thunderbird-default"
WM_ICON_NAME(STRING) = "Thunderbird"
_NET_WM_ICON_NAME(UTF8_STRING) = "Thunderbird"
WM_NAME(STRING) = "Thunderbird"
_NET_WM_NAME(UTF8_STRING) = "Thunderbird"
```

thunderbird regular context menu (with the issue):
```
WM_HINTS(WM_HINTS):
                Client accepts input or input focus: True
                Initial state is Normal State.
                window id # of group leader: 0x1200001
WM_WINDOW_ROLE(STRING) = "Popup"
_NET_WM_BYPASS_COMPOSITOR(CARDINAL) = 2
XdndAware(ATOM) = BITMAP
_NET_WM_OPAQUE_REGION(CARDINAL) = 
WM_TRANSIENT_FOR(WINDOW): window id # 0x1200068
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_POPUP_MENU
_NET_WM_SYNC_REQUEST_COUNTER(CARDINAL) = 18874718, 18874719
_NET_WM_USER_TIME(CARDINAL) = 1608771
_NET_WM_USER_TIME_WINDOW(WINDOW): window id # 0x120015d
WM_CLIENT_LEADER(WINDOW): window id # 0x1200001
_NET_WM_PID(CARDINAL) = 4145
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "e-mail"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified location: 0, 0
                program specified minimum size: 0 by 0
                program specified base size: 0 by 0
                window gravity: NorthWest
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING, _NET_WM_SYNC_REQUEST
WM_CLASS(STRING) = "Thunderbird", "thunderbird-default"
WM_ICON_NAME(STRING) = "Thunderbird"
_NET_WM_ICON_NAME(UTF8_STRING) = "Thunderbird"
WM_NAME(STRING) = "Thunderbird"
_NET_WM_NAME(UTF8_STRING) = "Thunderbird"
```

0*0 size seems to be a common property.

**Comment by 3hhh** [comment]
Hm... I just realized that I probably shouldn't have taken the xprops as I did above:
Qubes OS essentially runs two X servers in different VMs and the xprops above are from the VM where the applications are run, but not qtile.
qtile runs in another VM and may be presented different xprops by Qubes OS.

So here are the xprops from the VM qtile runs in:
[firefox_failing.txt](https://github.com/user-attachments/files/17810976/firefox_failing.txt)
[firefox_success.txt](https://github.com/user-attachments/files/17810977/firefox_success.txt)
[thunderbird_failing.txt](https://github.com/user-attachments/files/17810978/thunderbird_failing.txt)
[thunderbird_success.txt](https://github.com/user-attachments/files/17810979/thunderbird_success.txt)

Interesting points:
- Whenever there's a `ẀM_TRANSIENT_FOR`, there's an issue.
- Qubes OS strips most information, sometimes including the `WM_TRANSIENT_FOR` part. So this could still be a Qubes OS bug after all...

@marmarek : Can you possibly elaborate under which circumstances Qubes OS removes `WM_TRANSIENT_FOR` from windows and under which circumstances it passes it to the window manager?

**Comment by marmarek** [comment]
> @marmarek : Can you possibly elaborate under which circumstances Qubes OS removes `WM_TRANSIENT_FOR` from windows and under which circumstances it passes it to the window manager?

It's passed through if it points at a valid window at the moment when the window gets mapped by the application. If application changes it when the window is already mapped (as in - application first maps the window and only then sets WM_TRANSIENT_FOR), it's ignored.

**Comment by tych0** [comment]
Yeah, I am pretty suspicious of qtile's transient_for code, given there are (IMO, bugs) like https://github.com/tych0/qtile/commit/0c7f945f8dcba3e2a602bf73267fd285beb68822

I haven't had the time to understand the layering code in detail, but my guess is there's some problem there.

**Comment by elParaguayo** [comment]
> I haven't had the time to understand the layering code in detail, but my guess is there's some problem there.

I take no offence from that! I'm pretty sure that's where the issue is too.

**Comment by elParaguayo** [comment]
Yes, but there fact that our stacking code *doesn't* handle that window could be problem.

**Comment by tych0** [comment]
Yeah, my patch was just a guess, but I think the bug is likely in change_layer(). I am really having trouble understanding what this code is supposed to do, unfortunately.

It seems like https://github.com/qtile/qtile/blob/ea3f30b80c848659ba331c43d683125e588af095/libqtile/backend/x11/window.py#L1024-L1026

might be false for these windows, since they could be unmapped and wouldn't be in the current stack? Then we'd stop here and not actually do any work, which would trip this bug.

**Comment by 3hhh** [comment]
No, that's still not it.

I had [a look at the incoming X11 events](https://github.com/qtile/qtile/blob/16b870596e4087cfdaa21da63e1637ec2e5f6540/libqtile/backend/x11/core.py#L302) and noticed that qtile never receives any manage event. Ironically it receives an unmanage event, but it obviously ignores that. IIRC that also applies to other applications, not only firefox.

So essentially the bug happens as Xorg draws the window without qtile knowing of it. Fortunately it's on top once it's drawn first. However if the application decides to just hide/unmanage it instead of destroying it, it's z-axis position may change and upon unhide it's somewhere in between all other windows.

The remaining question is: Why is there no manage event received by qtile?

I only see a bunch of ConfigureNotify and PropertyNotify events.

**Comment by marmarek** [comment]
Isn't this exactly what override-redirect flag does? (It's X11 way for all kind of menus and notifications)

**Comment by elParaguayo** [comment]
Which is why I tried to do a PR that tracked (but didn't manage) those windows so we can at least ensure that our stacking code takes them into account. But there was obviously a flaw in my code somewhere!

**Comment by 3hhh** [comment]
Argh. I'm beginning to understand... So probably the context menu has an override-redirect flag, Xorg therefore doesn't forward it to qtile and qtile cannot position it on top of all other windows, if the application doesn't destroy the window, but just unmaps it for a while and later maps it again.

Imho it's not really a qtile bug as qtile intentionally doesn't handle those windows (it's the point of the override-redirect flag, is it not?).
The best it could _possibly_ do is to send a destroy after an unmap of unmanaged windows and the application would have to notice that and re-create the window.

It could still be a Qubes OS bug as Qubes OS [does a lot of override-redirect handling](https://github.com/QubesOS/qubes-gui-daemon/blob/main/gui-daemon/xside.c). From #4805 however it seems that not only Qubes OS is affected...

Also, it's probably not good application behaviour to re-use override-redirect windows, but it's unlikely that we'll be able to fix all the applications.

**Comment by marmarek** [comment]
On Wed, Nov 20, 2024 at 11:33:49AM -0800, 3hhh wrote:
> Also, it's probably not good application behaviour to re-use override-redirect windows, but it's unlikely that we'll be able to fix all the applications.

That's pretty common behavior - if you have a menu, usually the same
window gets reused for the same menu (so application doesn't need to
re-construct it each time).

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab

**Comment by tych0** [comment]
> Argh. I'm beginning to understand... So probably the context menu has an override-redirect flag, Xorg therefore doesn't forward it to qtile and qtile cannot position it on top of all other windows

But we *do* potentially rearrange the window's stacking in various places in `change_layer()`, so if we do e.g.
https://github.com/qtile/qtile/blob/6b240e5af6bca9058965e50c6b5295766b6fa781/libqtile/backend/x11/window.py#L1179-L1181
On an override-redirect window, the application won't necessarily know to re-stack it above again. I guess this is what #4583 is intending to fix.

**Comment by 3hhh** [comment]
Yes, if qtile somehow becomes aware of override-redirect windows, it can of course put its own windows below them.
Anyway #4583 didn't fix it for me neither.

**Comment by elParaguayo** [comment]
It's totally possible by that my man code in 4583 has some flaws in it. I think it's the right idea but may be poorly implemented.

**Comment by 3hhh** [comment]
I randmonly noticed that setting `bring_front_click = False` fixes 70% of the issue, i.e. that override-redirect windows stay on top of their parent.  

The remaining 30% it doesn't fix: They don't stay on top of all windows, i.e. an override-redirect window may be cut off at the border to other windows; this however is much less annoying than not seeing a window at all. #4805 and #4999 are probably about these 30%.

Override-redirect windows can always stay on top imho.

I believe what happens with `bring_front_click = True` is that the Xorg stacking order is modified on every click (windows are put above override-redirect windows) and on re-map they just show in that exact stacking location (unless they were newly created as e.g. `chromium` does it).

So in total this bug is just a [RTFM "bug"](https://docs.qtile.org/en/latest/manual/config/index.html) (btw the doc is quite excellent!).

It may however make sense to entirely remove the `bring_to_front_click` setting or at least put an even bigger footgun warning flag on it as from "When clicked, should the window be brought to the front or not." users like me tend to think "Yeah, sure, what else?." and set `bring_front_click = True`.
Yes, I had also read the comment about `kept_above`, but I don't have any such windows. Had it mentioned that it e.g. breaks `firefox` context menus, I'd not have set it.

**Comment by tych0** [comment]
I don't know, I still think it's a bug: bring_front_click should probably not fiddle with override-redirect stacking, since we're really not supposed to fiddle with those windows. If we fixed that, then everything would work like you expect.

**Comment by 3hhh** [comment]
I just noticed that the [dunst notification manager](https://github.com/dunst-project/dunst) also unmaps override-redirect windows and re-uses them at a later point in time. Those windows may appear behind other windows without override-redirect even with `bring_front_click = False`.

**Comment by 3hhh** [comment]
Still true, dear bot...

**Comment by tych0** [comment]
Does https://github.com/qtile/qtile/commit/67dae1971738ea5833694a1c24392f63c7e4079c work?

**Comment by 3hhh** [comment]
On 6/30/25 03:48, Tycho Andersen wrote:
> tych0 left a comment (qtile/qtile#5067)
> 
> Does https://github.com/qtile/qtile/commit/67dae1971738ea5833694a1c24392f63c7e4079c work?

Sorry, but no.

**Comment by tych0** [comment]
I have just been playing with this some more, and it turns out that *all* firefox menus are override-redirect:

https://github.com/mozilla-firefox/firefox/blob/fd531836ca5e48e18a5afc171418bcbcb9d268e2/widget/gtk/nsWindow.cpp#L6137-L6147

Qtile handles them accordingly, which is to say it doesn't touch them at all. But that means they are "sticky", so you can end up with floating windows. The easiest way to reproduce I have found is with the "do you want to share your location with this site" dialog. So go to e.g. https://browserleaks.com/geo and then switch to a different group. You'll see that the override-redirect geolocation popup is not automatically unmapped by firefox (because it's not a context menu, but a security question, so there's no default answer).

I don't really know if there's a fix here: the spec says we are explicitly supposed to ignore these windows. Firefox refuses to unmap it, because it doesn't want to default-answer a security question.

For other menus (right click, bookmarks, whatever) firefox correctly unmaps the override-redirect windows when it gets an UmapNotify from Qtile switching groups.

All this is to say: it seems like override-redirect is Working As Intended? With the possible exception of: on restart, if an override-redirect window is the window with x11 focus, I think things will probably break if that window is closed and the application doesn't reset focus somewhere else. But that would break more than just right clicks...

**Comment by tych0** [comment]
Also for future debugging: `xwininfo` does print the override redirect flag, whereas `xprop` does not.

**Comment by tych0** [comment]
Also also, note that we do not add override_redirect windows to qtile's window map at all. It is skipped by the logic here:

https://github.com/qtile/qtile/blob/4655d4a0542b94c5a59f8779ed7a0c2e2f33ad3d/libqtile/backend/x11/core.py#L769-L770

So I think the reality is: we are out-of-the loop for these windows almost completely. If they want to re-map themselves, then so be it, but they need to re-stack themselves if need be.


However, what you're reporting in this bug suggests we should potentially rethink our override_redirect behavior. Maybe we should just treat override-redirect window as "application set position" windows, but otherwise treat them normally. I wonder if that would fix things? Something like https://github.com/qtile/qtile/commit/55355f3c5dac9501c3e102a2681ece2cc70c2953 ?

In my testing, all the firefox windows behave exactly the same as they did before. But maybe that fixes things for other people, since qtile will "properly" reason about override-redirect windows since it's aware of them?

**Comment by 3hhh** [comment]
I _suspect_ the issue probably lies with the X11 stacking logic, which doesn't take override redirect into account, but re-stacks anyway using the X11 API leading to random stacking for override redirect windows.
qtile can do whatever re-stacking it wants, but it should always remain below the lowest override redirect window regardless of the map state and thus leave those in the order they were before. Since they are floating they may then appear in other parts of the screen than the parent window on e.g. window re-order, but that's an application issue.

An example:
```
Override-Redirect-Window A
Override-Redirect-Window B
Regular Window 1
Regular Window 2
```

Re-stacking `Regular Window 2` to the top should never result in:
```
Regular Window 2
Override-Redirect-Window A
Override-Redirect-Window B
Regular Window 1
```
(Probably that's what `bring_front_click = True` does and what layout switches may also trigger.)

It should rather always result in:
```
Override-Redirect-Window A
Override-Redirect-Window B
---- (there should be an invisible barrier for qtile here) ----
Regular Window 2
Regular Window 1
```

Of course it should also not re-stack the override redirect windows among each other in any way.

So qtile probably needs to take override-redirect windows into account for re-stacking / identify that barrier, but it doesn't need to care about them in any other place.

Anyway I guess that's why override-redirect is so much of a pain point about X11 and caused many bugs... Looking forward to switch to Wayland one day...

[!5397](https://github.com/qtile/qtile/pull/5397) may help with this, but I haven't tested it so far. Have you?

**Comment by tych0** [comment]
> qtile can do whatever re-stacking it wants, but it should always remain below the lowest override redirect window regardless of the map state and thus leave those in the order they were before.

Agreed, though without my patch above we don't track them at all.

> !5397 may help with this, but I haven't tested it so far. Have you?

I have tested the branch and seen no changes in behavior. However, I have not been able to reproduce the behavior you describe where we stack other things above them. I do not make extensive use of floating though, mostly my testing of that branch is to make sure there are no regressions, which I don't see any for my use cases.

**Comment by elParaguayo** [comment]
#5397  does nothing with override redirect windows as we don't manage them so they're invisible to the stacking logic.

**Comment by 3hhh** [comment]
> However, I have not been able to reproduce the behavior you describe where we stack other things above them. I do not make extensive use of floating though, mostly my testing of that branch is to make sure there are no regressions, which I don't see any for my use cases.

I see. I pretty much almost never use floating either though.

Anyway I just managed to reproduce it 5 out of 5 times on qtile 0.33 with the following approach:

1. On an empty workspace in MonadTall layout with `new_client_position = "bottom"`, open a terminal (Mod + Return for me).
2. Open firefox (I do it via Mod + r = run command).
4. Right click at the bottom of firefox to open the override redirect context menu. It should open to the top because there's not enough space below. Close it again by left clicking outside of it (like 1mm to the left of the context menu or so).
5. Open another terminal showing below firefox as per the MonadTall layout.
6. Click on the bottom of the firefox window again. The context menu should now start in the firefox window and be cut off at the terminal window below.

I'll test the proposed fix with that.

> #5397  does nothing with override redirect windows as we don't manage them so they're invisible to the stacking logic.

Yes, and imho that _may_ be an issue: In Xorg the top of the stack is always position 0 incl. override-redirect windows, in qtile it's without override-redirect windows, which would be Xorg position 2 = qtile position 0 in the above example. So if you re-stack using the Xorg API without _any_ knowledge of those override-redirect windows, you'll eventually get a Xorg stack position wrong. I may however not correctly have understood the old re-stack code (IIRC it asks Xorg for a list of windows and just filters override-redirect windows from that.)...

**Comment by 3hhh** [comment]
> I'll test the proposed fix with that.

Looks like I was right after all:

https://github.com/qtile/qtile/commit/55355f3c5dac9501c3e102a2681ece2cc70c2953 does nothing.

However initial testing with the aforementioned approach indicates that #5397 on top of 0.33 fixes this one.
I guess it may also fix the other Xorg issues related to override-redirect windows then.
So thanks to @elParaguayo for the PR!
I'll test it a bit more to see whether it causes other issues.

**Comment by elParaguayo** [comment]
I could probably modify that PR so that override redirect windows are added to a layet group above everything else so they're always at the top.

**Comment by 3hhh** [comment]
> I could probably modify that PR so that override redirect windows are added to a layet group above everything else so they're always at the top.

If it's not necessary, I'm against that.

Let's see whether I notice anything odd during the next few weeks.

**Comment by 3hhh** [comment]
Argh, please re-open, looks like my commit keyword auto-closed this, but I cannot re-open.

**Comment by 3hhh** [comment]
P.S.: I just tested the above test case with #5397 and `bring_front_click = True` and that appears to work correctly as well. :)

**Comment by 3hhh** [comment]
Still valid, closure depends on #5397.
